build_tools/scripts/integrate/auto_integrate.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/env python
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""Prototype of an LLVM auto-integrate script.
+
+WARNING: This script is a WIP that is being developed while stellaraccident@
+does LLVM integrates. If this warning is still present without activity after
+about Aug-2023, then consider it defunct.
+
+This script attempts to define a human-augmented workflow for staying up to
+date with LLVM. Roughly, it encourages the following flow:
+
+1. Integrator `start`'s a new integrate branch, which will reset llvm-project
+   to the next affecting change (currently defined as anything touching
+   MLIR but can be expanded).
+2. The integrator or automation periodically issues a `next` command, which
+   will advance LLVM to the next affecting commit (or an arbitrary future
+   commit).
+3. If the CI on the integrate branch signals a failure, then the integrator
+   should apply patches as appropriate to turn it green or manually decide
+   to advance anyway (i.e. around a breakage/revert).
```

Comment:
If this runs automated commit-by-commit, we may want to fine-tune the CI jobs that run while bumping, or rely on local builds/tests. We shouldn't run a full CI build (including benchmarks and all platforms, build systems, etc.) for every LLVM/MLIR commit, for example.

I could see a dedicated CI job that manages a bumping process:
1. pick an initial commit
2. build
  a. if successful, advance to the next commit (keeping all state like local caches / build directories, file downloads) then GOTO 2
  b. if unsuccessful, GOTO 3
3. signal to a developer

Reply:
Yeah, agreed. I was going to push more in that direction first, but it turned out that the script was called into service early to try to culprit find the A100 issue. I'll iterate on the whole setup more.